clippy_lints/src/legacy_integral_constants.rs

```diff
@@ -0,0 +1,147 @@
+
+impl<'tcx> LateLintPass<'tcx> for LegacyIntegralConstants {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'tcx>) {
+        if !self.msrv.meets(msrvs::STD_INTEGRAL_CONSTANTS) || in_external_macro(cx.sess(), expr.span) {
+            return;
+        }
+        let ExprKind::Path(qpath) = expr.kind else {
+            return;
+        };
+
+        // `std::<integer>::<CONST>` check
+        let (span, sugg, is_method) = if let QPath::Resolved(_, path) = qpath
+            && let Some(def_id) = path.res.opt_def_id()
+            && let Some(name) = path.segments.iter().last().map(|segment| segment.ident.name)
+            && let Some(module_name) = is_path_in_integral_module(cx, def_id)
+        {
+            (
+                expr.span,
+                format!("{module_name}::{name}"),
+                false,
+            )
+        // `<integer>::xxx_value` check
+        } else if let QPath::TypeRelative(ty, _) = qpath
+            && let TyKind::Path(ty_qpath) = ty.kind
+            && let Res::PrimTy(PrimTy::Int(_) | PrimTy::Uint(_)) = cx.qpath_res(&ty_qpath, ty.hir_id)
+            && let last_segment = last_path_segment(&qpath)
+            && let name = last_segment.ident.name.as_str()
+            && (name == "max_value" || name == "min_value")
+            // Also remove the `()`
+            && let Some(par_expr) = get_parent_expr(cx, expr)
+            && let ExprKind::Call(_, _) = par_expr.kind
+        {
+            (
+                qpath.last_segment_span().with_hi(par_expr.span.hi()),
+                name[..=2].to_ascii_uppercase(),
+                true,
+            )
+        } else {
+            return;
+        };
+
+        if !is_from_proc_macro(cx, expr) {
+            let msg = if is_method {
+                "usage of a legacy integral constant method"
+            } else {
+                "usage of a legacy integral constant"
+            };
+    
+            span_lint_and_sugg(
+                cx,
+                LEGACY_INTEGRAL_CONSTANTS,
+                span,
+                msg,
+                "try using the associated constant instead",
+                sugg,
+                Applicability::MachineApplicable,
+            );
+        }
+    }
+
+    extract_msrv_attr!(LateContext);
+}
+
+fn is_path_in_integral_module(cx: &LateContext<'_>, def_id: DefId) -> Option<Symbol> {
+    macro_rules! is_path_in_integral_module {
+        ($segments:ident, $($name:ident),+$(,)?) => {
+            $(
+                ::clippy_utils::paths::$name.iter()
+                    .map(|i| ::rustc_span::Symbol::intern(i))
+                    .eq($segments.iter().copied())
+            )||+
+        };
+    }
+
+    // This is awful, but there isn't really much of a better way to do it unfortunately.
+    // FIXME: If possible, let's simplify this.
+    if let [segments @ .., _] = &*cx.get_def_path(def_id)
+        && is_path_in_integral_module![
```

Comment:
Does this work?

```rust
if let [sym::core, sym::u8 | sym::i8 | sym::u16 | etc., last] = .. {
    Some(last)
} else {
    None
}
```

Reply:
Hmm potentially so. I'll see if I can get that to work later

Reply:
Yep it does, thank you!